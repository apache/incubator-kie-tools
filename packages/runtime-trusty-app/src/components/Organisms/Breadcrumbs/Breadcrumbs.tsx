/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from "react";
import useBreadcrumbs from "use-react-router-breadcrumbs";
import { Breadcrumb, BreadcrumbItem } from "@patternfly/react-core/dist/js/components";
import { RouteComponentProps } from "react-router-dom";
import ExecutionId from "../../Atoms/ExecutionId/ExecutionId";
import TrustyLink from "../../Atoms/TrustyLink/TrustyLink";

const Breadcrumbs = () => {
  const breadcrumbs = useBreadcrumbs(routes, { excludePaths });

  return (
    <>
      {breadcrumbs.length >= 2 && (
        <Breadcrumb>
          {breadcrumbs.map(({ match, location, breadcrumb }) => {
            return (
              <BreadcrumbItem
                className="breadcrumb-item"
                key={match.url}
                isActive={location.pathname === match.url}
                render={({ className }) => (
                  <TrustyLink url={match.url} className={className}>
                    {breadcrumb}
                  </TrustyLink>
                )}
              />
            );
          })}
        </Breadcrumb>
      )}
    </>
  );
};

type executionIdParam = {
  id: string;
};

interface AuditDetailBreadcrumbProps extends RouteComponentProps<executionIdParam> {}

const AuditDetailBreadcrumb = (props: AuditDetailBreadcrumbProps) => {
  const { match } = props;
  return (
    <span>
      Execution <ExecutionId id={match.params.id} />
    </span>
  );
};

const routes = [
  { path: "/audit", breadcrumb: "Audit investigation" },
  {
    path: "/audit/:executionType/:id/outcomes-details",
    breadcrumb: "Outcomes details",
  },
  // the following route is needed to display a dedicated breadcrumb path for executions with only 1 outcome
  {
    path: "/audit/:executionType/:id/single-outcome",
    breadcrumb: "Outcome",
  },
  {
    path: "/audit/:executionType/:id/model-lookup",
    breadcrumb: "Model lookup",
  },
  {
    path: "/audit/:executionType/:id/input-data",
    breadcrumb: "Input data",
  },
  {
    path: "/audit/:executionType/:id/counterfactual-analysis",
    breadcrumb: "Counterfactual analysis",
  },
  {
    path: "/audit/:executionType/:id",
    breadcrumb: AuditDetailBreadcrumb,
  },
];

const excludePaths = ["/", "/audit/:executionType"];

export default Breadcrumbs;
