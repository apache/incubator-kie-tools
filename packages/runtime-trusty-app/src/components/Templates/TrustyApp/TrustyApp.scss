.trusty-logo {
  width: 158px;
}

/*
  Fixing weird toolbar extra spacing introduced after updating to PF 2021.08 (see FAI-611)
  The issue is not present in more recent versions of PF.
  To be removed after the next PF update on kogito-apps.
 */
.pf-c-toolbar__content[hidden] {
  display: none;
}
