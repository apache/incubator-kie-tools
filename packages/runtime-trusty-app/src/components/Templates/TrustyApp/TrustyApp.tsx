/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React, { useContext, useState } from "react";
import { BrowserRouter, NavLink, Redirect, Route, Switch, useLocation } from "react-router-dom";
import {
  Avatar,
  Brand,
  Nav,
  NavItem,
  NavList,
  Page,
  PageHeader,
  PageHeaderTools,
  PageSidebar,
} from "@patternfly/react-core/dist/js/components";
import AuditOverview from "../AuditOverview/AuditOverview";
import kogitoLogo from "../../../static/images/kogitoLogo.svg";
import AuditDetail from "../AuditDetail/AuditDetail";
import imgAvatar from "../../../static/images/user.svg";
import Breadcrumbs from "../../Organisms/Breadcrumbs/Breadcrumbs";
import NotFound from "../NotFound/NotFound";
import ApplicationError from "../ApplicationError/ApplicationError";
import { TrustyContextValue } from "../../../types";
import { datePickerSetup } from "../../Molecules/DatePicker/DatePicker";
// manually importing Red Hat Mono font because not yet provided by PatternFly
// see: https://github.com/patternfly/patternfly/issues/4021
// when updating PF to a version that will include it, it will be
// important to remove this duplication
import "../../../static/fonts/RedHatMono/RedHatMono.css";
import "./TrustyApp.scss";

datePickerSetup();

type TrustyAppProps = {
  /** Enable counterfactual analysis feature */
  counterfactualEnabled: boolean;
  /** Enable explainability information inside decisions */
  explanationEnabled: boolean;
  /** Configuration options to embed Trusty inside a parent application  */
  containerConfiguration?: ContainerConfigurationProps;
};

type ContainerConfigurationProps = {
  /** Include the page layout wrapper with sidebar navigation and breadcrumbs */
  pageWrapper?: boolean;
  /** Use an optional Server root for all REST operations */
  serverRoot: string;
  /** Use an optional base path for internal routes */
  basePath?: string;
  /**
   * Do not include a rect router (BrowserRouter) within TrustyApp.
   * The host application will have to include one in order to make
   * Trusty internal Routes work. */
  excludeReactRouter?: boolean;
  /**
   * Use traditional anchor links with href attributes. When set to false
   * all links will be handled with a onClick handler pushing url
   * to the browser history */
  useHrefLinks?: boolean;
};

const TrustyApp: React.FC<TrustyAppProps> = (props) => {
  const baseUrl = window.TRUSTY_ENDPOINT || process.env.KOGITO_TRUSTY_API_HTTP_URL!;

  const {
    counterfactualEnabled,
    explanationEnabled,
    containerConfiguration: {
      pageWrapper = true,
      serverRoot = baseUrl,
      basePath = "",
      excludeReactRouter = false,
      useHrefLinks = true,
    } = {},
  } = props;

  return (
    <TrustyContext.Provider
      value={{
        config: {
          counterfactualEnabled,
          explanationEnabled,
          serverRoot: serverRoot,
          basePath: basePath,
          useHrefLinks,
        },
      }}
    >
      <>
        {excludeReactRouter ? (
          <TrustyAppContainer pageWrapper={pageWrapper} />
        ) : (
          <BrowserRouter>
            <TrustyAppContainer pageWrapper={pageWrapper} />
          </BrowserRouter>
        )}
      </>
    </TrustyContext.Provider>
  );
};

export default TrustyApp;

export const TrustyContext = React.createContext<TrustyContextValue>({} as TrustyContextValue);

type TrustyAppContainerProps = {
  pageWrapper: boolean;
};

const TrustyAppContainer = ({ pageWrapper }: TrustyAppContainerProps) => {
  const location = useLocation();
  const [isMobileView, setIsMobileView] = useState(false);
  const [isNavOpenDesktop, setIsNavOpenDesktop] = useState(true);
  const [isNavOpenMobile, setIsNavOpenMobile] = useState(false);
  const { config } = useContext(TrustyContext);

  const onNavToggleDesktop = () => {
    setIsNavOpenDesktop(!isNavOpenDesktop);
  };

  const onNavToggleMobile = () => {
    setIsNavOpenMobile(!isNavOpenMobile);
  };
  const handlePageResize = (props: { windowSize: number; mobileView: boolean }) => {
    // closing sidebar menu when resolution is < 1200
    if (props.windowSize < 1200) {
      if (!isMobileView) setIsMobileView(true);
    } else {
      if (isMobileView) setIsMobileView(false);
    }
  };

  const PageNav = (
    <Nav aria-label="Nav" theme="dark">
      <NavList>
        <NavItem isActive={location.pathname.startsWith("/audit")} ouiaId="audit-item">
          <NavLink to="/audit">Audit investigation</NavLink>
        </NavItem>
      </NavList>
    </Nav>
  );

  const Sidebar = (
    <PageSidebar nav={PageNav} isNavOpen={isMobileView ? isNavOpenMobile : isNavOpenDesktop} theme="dark" />
  );

  const Header = (
    <PageHeader
      logo={<Brand src={kogitoLogo} alt="Kogito TrustyAI" className="trusty-logo" />}
      logoProps={{ href: `${config.basePath}/`, className: "trusty-home-link" }}
      headerTools={
        <PageHeaderTools>
          <Avatar src={imgAvatar} alt="Avatar image" />
        </PageHeaderTools>
      }
      showNavToggle={isMobileView}
      onNavToggle={isMobileView ? onNavToggleMobile : onNavToggleDesktop}
      isNavOpen={isMobileView ? isNavOpenMobile : isNavOpenDesktop}
    />
  );

  const Routes = (
    <Switch>
      <Route exact path={`${config.basePath}/`}>
        <Redirect to={`${config.basePath}/audit`} />
      </Route>
      <Route exact path={`${config.basePath}/audit`}>
        <AuditOverview />
      </Route>
      <Route path={`${config.basePath}/audit/:executionType/:executionId`}>
        <AuditDetail />
      </Route>
      <Route exact path="/error">
        <ApplicationError />
      </Route>
      <Route path="/not-found" component={NotFound} />
      <Redirect to="/not-found" />
    </Switch>
  );

  return (
    <>
      {pageWrapper ? (
        <Page header={Header} sidebar={Sidebar} breadcrumb={<Breadcrumbs />} onPageResize={handlePageResize}>
          {Routes}
        </Page>
      ) : (
        <>{Routes}</>
      )}
    </>
  );
};
