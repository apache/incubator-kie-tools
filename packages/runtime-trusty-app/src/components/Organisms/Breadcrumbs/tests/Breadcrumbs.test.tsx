/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from "react";
import Breadcrumbs from "../Breadcrumbs";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { MemoryRouterProps } from "react-router";
import { TrustyContext } from "../../../Templates/TrustyApp/TrustyApp";

const setupWrapper = (routerEntries: MemoryRouterProps["initialEntries"]) => {
  return mount(
    <MemoryRouter initialEntries={routerEntries}>
      <TrustyContext.Provider
        value={{
          config: {
            counterfactualEnabled: true,
            explanationEnabled: true,
            basePath: "",
            useHrefLinks: true,
          },
        }}
      >
        <Breadcrumbs />
      </TrustyContext.Provider>
    </MemoryRouter>
  );
};

describe("Breadcrumbs", () => {
  test("renders correctly", () => {
    const wrapper = setupWrapper(["/audit"]);
    const breadcrumbs = wrapper.find(Breadcrumbs);

    expect(breadcrumbs).toMatchSnapshot();
    expect(breadcrumbs.find("li.breadcrumb-item")).toHaveLength(0);
  });

  test("renders outcome details breadcrumbs links", () => {
    const executionId = "b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000";
    const wrapper = setupWrapper([
      {
        pathname: `/audit/decision/${executionId}/outcomes`,
        key: "execution",
      },
    ]);
    const breadcrumbs = wrapper.find(Breadcrumbs);

    expect(breadcrumbs).toMatchSnapshot();
    expect(breadcrumbs.find("li.breadcrumb-item")).toHaveLength(3);
    expect(breadcrumbs.find("li.breadcrumb-item").at(0).text()).toMatch("Audit investigation");
    expect(breadcrumbs.find("li.breadcrumb-item").at(1).text()).toMatch(`Execution #${executionId.substring(0, 8)}`);
    expect(breadcrumbs.find("li.breadcrumb-item").at(2).text()).toMatch("Outcomes");
    expect(breadcrumbs.find("BreadcrumbItem").at(0).prop("isActive") as boolean).toBeFalsy();
    expect(breadcrumbs.find("BreadcrumbItem").at(1).prop("isActive") as boolean).toBeFalsy();
    expect(breadcrumbs.find("BreadcrumbItem").at(2).prop("isActive") as boolean).toBeTruthy();
  });
});
