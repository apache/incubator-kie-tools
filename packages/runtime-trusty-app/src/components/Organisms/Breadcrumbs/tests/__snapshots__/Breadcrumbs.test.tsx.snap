// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumbs renders correctly 1`] = `<Breadcrumbs />`;

exports[`Breadcrumbs renders outcome details breadcrumbs links 1`] = `
<Breadcrumbs>
  <Breadcrumb>
    <nav
      aria-label="Breadcrumb"
      className="pf-c-breadcrumb"
      data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
      data-ouia-component-type="PF4/Breadcrumb"
      data-ouia-safe={true}
    >
      <ol
        className="pf-c-breadcrumb__list"
      >
        <BreadcrumbItem
          className="breadcrumb-item"
          isActive={false}
          key=".$/audit"
          render={[Function]}
          showDivider={false}
        >
          <li
            className="pf-c-breadcrumb__item breadcrumb-item"
          >
            <TrustyLink
              className="pf-c-breadcrumb__link"
              url="/audit"
            >
              <Link
                className="pf-c-breadcrumb__link"
                to="/audit"
              >
                <LinkAnchor
                  className="pf-c-breadcrumb__link"
                  href="/audit"
                  navigate={[Function]}
                >
                  <a
                    className="pf-c-breadcrumb__link"
                    href="/audit"
                    onClick={[Function]}
                  >
                    <span
                      key="/audit"
                    >
                      Audit investigation
                    </span>
                  </a>
                </LinkAnchor>
              </Link>
            </TrustyLink>
          </li>
        </BreadcrumbItem>
        <BreadcrumbItem
          className="breadcrumb-item"
          isActive={false}
          key=".$/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000"
          render={[Function]}
          showDivider={true}
        >
          <li
            className="pf-c-breadcrumb__item breadcrumb-item"
          >
            <span
              className="pf-c-breadcrumb__item-divider"
            >
              <AngleRightIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style={
                    Object {
                      "verticalAlign": "-0.125em",
                    }
                  }
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </AngleRightIcon>
            </span>
            <TrustyLink
              className="pf-c-breadcrumb__link"
              url="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000"
            >
              <Link
                className="pf-c-breadcrumb__link"
                to="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000"
              >
                <LinkAnchor
                  className="pf-c-breadcrumb__link"
                  href="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000"
                  navigate={[Function]}
                >
                  <a
                    className="pf-c-breadcrumb__link"
                    href="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000"
                    onClick={[Function]}
                  >
                    <AuditDetailBreadcrumb
                      key="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000"
                      location={
                        Object {
                          "hash": "",
                          "key": "execution",
                          "pathname": "/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000/outcomes",
                          "search": "",
                        }
                      }
                      match={
                        Object {
                          "isExact": true,
                          "params": Object {
                            "executionType": "decision",
                            "id": "b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000",
                          },
                          "path": "/audit/:executionType/:id",
                          "url": "/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000",
                        }
                      }
                    >
                      <span>
                        Execution 
                        <ExecutionId
                          id="b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000"
                        >
                          <span
                            className="execution-id"
                          >
                            #
                            b2b0ed8d
                          </span>
                        </ExecutionId>
                      </span>
                    </AuditDetailBreadcrumb>
                  </a>
                </LinkAnchor>
              </Link>
            </TrustyLink>
          </li>
        </BreadcrumbItem>
        <BreadcrumbItem
          className="breadcrumb-item"
          isActive={true}
          key=".$/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000/outcomes"
          render={[Function]}
          showDivider={true}
        >
          <li
            className="pf-c-breadcrumb__item breadcrumb-item"
          >
            <span
              className="pf-c-breadcrumb__item-divider"
            >
              <AngleRightIcon
                color="currentColor"
                noVerticalAlign={false}
                size="sm"
              >
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style={
                    Object {
                      "verticalAlign": "-0.125em",
                    }
                  }
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </AngleRightIcon>
            </span>
            <TrustyLink
              className="pf-c-breadcrumb__link pf-m-current"
              url="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000/outcomes"
            >
              <Link
                className="pf-c-breadcrumb__link pf-m-current"
                to="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000/outcomes"
              >
                <LinkAnchor
                  className="pf-c-breadcrumb__link pf-m-current"
                  href="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000/outcomes"
                  navigate={[Function]}
                >
                  <a
                    className="pf-c-breadcrumb__link pf-m-current"
                    href="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000/outcomes"
                    onClick={[Function]}
                  >
                    <span
                      key="/audit/decision/b2b0ed8d-c1e2-46b5-3ac54ff4beae-1000/outcomes"
                    >
                      Outcomes
                    </span>
                  </a>
                </LinkAnchor>
              </Link>
            </TrustyLink>
          </li>
        </BreadcrumbItem>
      </ol>
    </nav>
  </Breadcrumb>
</Breadcrumbs>
`;
